File: SOC_Particle/src/Coulombs.cpp
//
// MIT License
//
// Copyright (C) 2026 - Dave Gutz
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "application.h"
#include <math.h>
#include "Battery.h"
#include "Coulombs.h"
#include "parameters.h"
#include "command.h"

#include "subs.h" // delay_no_block

extern SavedPars sp;      // Various parameters to be static at system level and saved through power cycle
extern VolatilePars ap; // Various adjustment parameters shared at system level
extern CommandPars cp;    // Various parameters to be static at system level
extern PublishPars pp;    // For publishing


// class Coulombs
Coulombs::Coulombs() {}
Coulombs::Coulombs(double *sp_delta_q, const double q_cap_rated,
  const double s_coul_eff, const float dx_voc, const float dy_voc, const float dz_voc)
  : resetting_(false), d_delta_q_(0.), delta_q_abs_(0.), delta_q_inf_(0.), delta_q_neg_(0.), delta_q_pos_(0.), dt_(0.),
    q_(q_cap_rated), q_capacity_(q_cap_rated), q_cap_rated_(q_cap_rated), q_cap_rated_scaled_(q_cap_rated), q_inf_(0.), q_min_(0.),
    sat_(true), saturated_(false), soc_(1.), soc_ekf_min_(0.), soc_inf_(0.), soc_min_(0.), sp_delta_q_(sp_delta_q),
    tb_f_(0.), tb_f_rate_(0.), Tb_f_(0.), Tb_f_rate_(0.), time_neg_(0.), time_pos_(0.), chem_()
    {
      coul_eff_ = chem_.coul_eff*s_coul_eff;
      soc_ekf_min_ = chem_.soc_ekf_min;
      put_dx_voc(dx_voc);
      put_dy_voc(dy_voc);
      put_dz_voc(dz_voc);
    }
Coulombs::~Coulombs() {}


// operators
// Pretty print
void Coulombs::pretty_print()
{
#ifndef SOFT_DEPLOY_PHOTON
  Serial.printf("Coulombs:\n");
  Serial.printf(" coul_eff%9.5f\n", coul_eff_);
  Serial.printf(" d_delta_q%9.1f, C\n", d_delta_q_);
  Serial.printf(" delta_q%9.1f, C\n", *sp_delta_q_);
  Serial.printf(" delta_q_inf/delta_q_abs%9.1f / %9.1f %8.4f C\n", delta_q_inf_, delta_q_abs(), delta_q_inf_/delta_q_abs());
  Serial.printf(" delta_q_neg%9.1f C, time_neg%9.1f s\n", delta_q_neg_, time_neg_);
  Serial.printf(" delta_q_pos%9.1f C, time_pos%9.1f s\n", delta_q_pos_, time_pos_);
  Serial.printf(" dt%9.6f, s\n", dt_);
  Serial.printf(" mod_code %d\n", mod_code());
  Serial.printf(" mod %s\n", chem_.decode(mod_code()).c_str());
  Serial.printf(" q%9.1f, C\n", q_);
  Serial.printf(" q_cap%9.1f, C\n", q_capacity_);
  Serial.printf(" q_cap_rat%9.1f, C\n", q_cap_rated_);
  Serial.printf(" q_cap_rat_scl%9.1f, C\n", q_cap_rated_scaled_);
  Serial.printf(" q_min%9.1f, C\n", q_min_);
  Serial.printf(" resetting %d\n", resetting_);
  Serial.printf(" sat %d\n", sat_);
  Serial.printf(" soc%8.4f\n", soc_);
  Serial.printf(" soc_inf%8.4f\n", soc_inf_);
  Serial.printf(" soc_min%8.4f\n", soc_min_);
  Serial.printf(" tb_f_%5.1f dg C\n", tb_f_);
  Serial.printf(" rated_t%5.1f dg C\n", chem_.rated_temp);
  Serial.printf(" tb_f_rate%9.5f dg C / s\n", tb_f_rate_);
  Serial.printf(" tb_f_rate%9.5f dg C / s\n", Tb_f_rate_);
  Serial.printf("Coulombs (mod_code=%d) ", mod_code());
  Serial.printf("Coulombs: silent DEPLOY\n");
  Serial.printf(" Chemistry::\n");
  chem_pretty_print();
#endif
}

// functions

// Scale size of battery and adjust as needed to preserve delta_q.  Tb unchanged.
// Goal is to scale battery and see no change in delta_q on screen of 
// test comparisons.   The rationale for this is that the battery is frequently saturated which
// resets all the model parameters.   This happens daily.   Then both the model and the battery
// are discharged by the same current so the delta_q will be the same.
void Coulombs::apply_cap_scale(const double scale)
{
  q_cap_rated_scaled_ = scale * q_cap_rated_;
  q_capacity_ = calculate_capacity(tb_f_);  // tb_f_ usually Tb_f to reduce electrical noise effects
  q_ = *sp_delta_q_ + q_capacity_; // preserve delta_q, deficit since last saturation (like real life)
  soc_ = q_ / q_capacity_;
  resetting_ = true;     // momentarily turn off saturation check
}

// Memory set, adjust book-keeping as needed.  delta_q, capacity, temp preserved
void Coulombs::apply_delta_q(const double delta_q)
{
  *sp_delta_q_ = delta_q;
  q_ = *sp_delta_q_ + q_capacity_;
  soc_ = q_ / q_capacity_;
  resetting_ = true;     // momentarily turn off saturation check
}

// Memory set, adjust book-keeping as needed.  q_cap_ etc presesrved
void Coulombs::apply_delta_q_t(const bool reset)
{
  if ( !reset ) return;
  q_capacity_ = calculate_capacity(tb_f_);
  q_ = q_capacity_ + *sp_delta_q_;
  soc_ = q_ / q_capacity_;
  resetting_ = true;
}
void Coulombs::apply_delta_q_t(const double delta_q, const double tb_f)
{
  tb_f_ = tb_f;
  *sp_delta_q_ = delta_q;
  apply_delta_q_t(true);
}


// Memory set, adjust book-keeping as needed.  delta_q preserved
void Coulombs::apply_soc(const double soc, const double tb_f)
{
  soc_ = soc;
  q_capacity_ = calculate_capacity(tb_f);
  q_ = soc*q_capacity_;
  *sp_delta_q_ = q_ - q_capacity_;
  resetting_ = true;     // momentarily turn off saturation check
}

// Capacity
double Coulombs::calculate_capacity(const double tb_f)
{
  return( q_cap_rated_scaled_ * (1. + chem_.dqdt*(tb_f - chem_.rated_temp)) );
}

/* Coulombs::count_coulombs:  Count coulombs based on true=actual capacity
Inputs:
  dt              Integration step, s
  reset_temp      Temperature frame reset flag, T=resetting
  tb_f            Battery temperature lagged and rate-limited, deg C
  charge_curr     Charge, A
  sat             Indication that battery is saturated, T=saturated
  coul_eff        Coulombic efficiency - the fraction of charging input that gets turned into usable Coulombs
Outputs:
  q_capacity_     Saturation charge at temperature, C
  *sp_delta_q_    Charge change since saturated, C
  resetting_      Sticky flag for initialization, T=reset
  soc_            Fraction of saturation charge (q_capacity_) available (0-1) 
  soc_min_        Estimated soc where battery BMS will shutoff current, fraction
  q_min_          Estimated charge at low voltage shutdown, C\
*/
double Coulombs::count_coulombs(Sensors *Sen, const bool reset_temp, const float charge_curr, const bool sat,
  const bool saturated)
{
    // Inputs
    dt_ = Sen->T();
    tb_f_ = Sen->Tb_f();
    tb_f_rate_ = Sen->Tb_f_rate();
    Tb_f_ = Sen->Tb_f();
    Tb_f_rate_ = Sen->Tb_f_rate();
    d_delta_q_ = charge_curr * dt_;

    // State change
    double d_delta_q_inf = d_delta_q_;
    if ( charge_curr>0. && !sp.tweak_test() ) d_delta_q_ *= coul_eff_;
    // Capacity changes withi temperature so this effect would be double if used
    // d_delta_q_ -= chem_.dqdt*q_capacity_*tb_f_rate_*dt_;
    d_delta_q_inf = d_delta_q_;
    sat_ = sat;
    saturated_ = saturated;

    // Saturation.   Goal is to set q_capacity and hold it so remember last saturation status.
    if ( saturated_ )
    {
        if ( d_delta_q_ > 0 )
        {
            d_delta_q_ = 0.;
            if ( !resetting_ )
            {
              *sp_delta_q_ = 0.;
            }
        }
        else if ( reset_temp )
        {
          *sp_delta_q_ = 0.;
        }
    }
    // else if ( reset_temp && !ap.fake_faults() ) *sp_delta_q_ = delta_q_ekf;  // Solution to booting up unsaturated
    resetting_ = false;     // one pass flag

    // Integration.   Can go to negative
    q_capacity_ = calculate_capacity(tb_f_);

    // soc integrator: gated only by reset_temp.  cp.inf_reset (HR) MUST NOT touch *sp_delta_q_,
    // otherwise HR would slam soc to 1.0 — see commit history.  Per-window pos/neg/time accumulators
    // are stepped here too; if cp.inf_reset is also asserted this cycle, the inf-counter block
    // below will zero them right after, so the net effect is correct.
    if ( !reset_temp )
    {
      *sp_delta_q_ = max(min(*sp_delta_q_ + d_delta_q_, 0.0), -q_capacity_*1.5);
      if ( d_delta_q_ > 0. )
      {
        delta_q_pos_ += d_delta_q_;
        time_pos_ += dt_;
      }
      else
      {
        delta_q_neg_ += d_delta_q_;
        time_neg_ += dt_;
      }
    }

    // History (inf-counter) family: cp.inf_reset (HR) zeroes them so soc_inf -> 1.0; reset_temp
    // rebaselines them to current *sp_delta_q_.  Neither path disturbs *sp_delta_q_ itself.
    if ( cp.inf_reset )
    {
      delta_q_inf_ = 0.;
      delta_q_abs_ = 0.;
      delta_q_pos_ = 0.;
      delta_q_neg_ = 0.;
      time_pos_ = 0.;
      time_neg_ = 0.;
      cp.inf_reset = false;
    }
    else if ( reset_temp )
    {
      delta_q_abs_ = *sp_delta_q_ / 2.;
      delta_q_inf_ = *sp_delta_q_;
      delta_q_neg_ = *sp_delta_q_;
      delta_q_pos_ = 0.;
      time_pos_ = 0.;
      time_neg_ = 0.;
    }
    else
    {
      delta_q_inf_ += d_delta_q_inf;
      delta_q_abs_ += abs(d_delta_q_inf) / 2.;
    }

    q_ = q_capacity_ + *sp_delta_q_;
    q_inf_ = q_capacity_ + delta_q_inf_;

    // Normalize
    soc_ = q_ / q_capacity_;
    soc_inf_ = q_inf_ / q_capacity_;
    soc_min_ = chem_.soc_min_T_->interp(tb_f_);
    q_min_ = soc_min_ * q_capacity_;

    // Save and return
    
    if ( sp.debug()==36 )
      sendTxBuf(String::format("BM::CC: cc %7.3f dt%9.6f dq_T%9.2f, coul_eff%7.3f d_delta_q%9.2f sp_delta_q_%9.2f q%9.2f\n",
        charge_curr, dt_, -chem_.dqdt*q_capacity_*Tb_f_rate_*dt_, coul_eff_, d_delta_q_, *sp_delta_q_, q_), true, IN_SERVICE);

    if ( sp.debug()==-99 )
      sendTxBuf(String::format("sat, dt_, tb_f_, charge_curr, dq, dqt+, ddq, q, soc_min soc, %d, %7.4f,%7.4f,%7.4f,%7.4f,%7.4f,%7.4f,%12.1f,%10.7f,%10.7f,\n",
        sat, dt_, tb_f_, charge_curr, charge_curr * dt_, chem_.dqdt*q_capacity_*Tb_f_rate_*dt_, d_delta_q_, q_, soc_min_, soc_),
        true, IN_SERVICE);

    return ( soc_ );
}

// Prevent overflows
double nice_zero(const double in, const double thr)
{
    double out = thr;
    if ( abs(in) < thr )
    {
      if ( in < 0. ) out = -thr; 
    }
    else
      out = in;
    return (out);
}

float nice_zero(const float in, const float thr)
{
    float out = thr;
    if ( abs(in) < thr )
    {
      if ( in < 0. ) out = -thr; 
    }
    else
      out = in;
    return (out);
}

File: SOC_Particle/pyStateOfCharge/Coulombs.py
# BatteryEKF - general purpose battery class for embedded EKF
# Copyright (C) 2026 Dave Gutz
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# See http://www.fsf.org/licensing/licenses/lgpl.txt for full license text.

"""Coulomb Counting.  1 Ampere is 1 Coulombs/second charge rate of change.   Herein are various methods to keep track
of the totals and standardize the calculations."""

# Constants
from Chemistry_BMS import Chemistry
import Globals as G


class Coulombs:
    """Coulomb Counting"""

    def __init__(self, OPT=None, q_cap_rated=None, q_cap_rated_scaled=None, t_rated=None, tweak_test=False, mod_code=0,
                 dvoc=0., Dw=0.):
        if OPT is not None and hasattr(OPT, 'unit'):
            unit = OPT.unit
        else:
            unit = ''
        self.q_cap_rated = q_cap_rated
        self.q_cap_rated_scaled = q_cap_rated_scaled
        self.t_rated = t_rated
        self.delta_q = 0.
        self.d_delta_q = 0.
        self.q = 0.
        self.q_capacity = 0.
        self.soc_min = 0.
        self.soc = 1.
        self.resetting = True
        self.q_min = 0.
        self.sat = True
        self.saturated = True
        self.chm = mod_code
        self.tweak_test = tweak_test
        self.reset = False
        self.Tb_f = 0.
        self.chemistry = Chemistry(mod_code=mod_code, dvoc=dvoc, unit=unit, Dw=Dw)
        self.chemistry.assign_all_mod(mod_code, unit=unit)

    def __str__(self, prefix=''):
        """Returns representation of the object"""
        s = prefix + "Coulombs:\n"
        s += "  q_cap_rated = {:9.1f}    // Rated capacity at t_rated_, saved for future scaling, C\n"\
            .format(self.q_cap_rated)
        s += "  q_cap_rated_scaled = {:9.1f} // Applied rated capacity at t_rated_, after scaling, C\n"\
            .format(self.q_cap_rated_scaled)
        s += "  q_capacity = {:9.1f}     // Saturation charge at temperature, C\n".format(self. q_capacity)
        s += "  q =          {:9.1f}     // Present charge available to use, C\n".format(self. q)
        s += "  delta_q      {:9.1f}     // Charge since saturated, C\n".format(self. delta_q)
        s += "  soc =        {:7.3f}       // Fraction of saturation charge (q_capacity_) available (0-1)  soc)\n"\
            .format(self.soc)
        s += "  sat =          {:1.0f}          // Indication from caller that battery is saturated, T=saturated\n"\
            .format(self.sat)
        s += "  t_rated =    {:5.1f}         // Rated temperature, deg C\n".format(self.t_rated)
        s += "  Tb_f    =    {:5.1f}         // Rated temperature, deg C\n".format(self.tb_f)
        # s += "  temp_rlim =     {:7.3f}       // Tbatt rate limit, deg C / s\n".format(self.temp_rlim)
        s += "  resetting =     {:d}          // Flag to test coulomb counters, T = external reset of counter\n"\
            .format(self.resetting)
        s += "  soc_min =    {:7.3f}       // Lowest soc for power delivery.   Arises with temp < 20 C\n"\
            .format(self.soc_min)
        s += "  tweak_test =    {:d}          // Driving signal injection completely using software inj_soft_bias\n"\
            .format(self.tweak_test)
        s += "  coul_eff =   {:8.4f}      // Coulombic efficiency- fraction of charging turned into usable Coulombs\n".\
            format(self.chemistry.coul_eff)
        return s

    def apply_cap_scale(self, scale):
        """ Scale size of battery and adjust as needed to preserve delta_q.  Tb unchanged.
        Goal is to scale battery and see no change in delta_q on screen of
        test comparisons.   The rationale for this is that the battery is frequently saturated which
        resets all the model parameters.   This happens daily.   Then both the model and the battery
        are discharged by the same current so the delta_q will be the same."""
        self.q_cap_rated_scaled = scale * self.q_cap_rated
        self.q_capacity = self.calculate_capacity(tb_f=self.Tb_f)
        self.q = self.delta_q + self.q_capacity  # preserve self.delta_q, deficit since last saturation(like real life)
        if self.q_capacity != 0. and self.q_capacity is not None:
            self.soc = self.q / self.q_capacity
        self.resetting = True  # momentarily turn off saturation check

    # Memory set, adjust bookkeeping as needed.delta_q, capacity, temp preserved void
    def apply_delta_q_brief(self, delta_q):
        self.delta_q = delta_q
        self.q = self.delta_q + self.q_capacity
        if self.q_capacity != 0. and self.q_capacity is not None:
            self.soc = self.q / self.q_capacity
        self.resetting = True  # momentarily turn off saturation check

    def apply_delta_q_t(self, delta_q, tb_f):
        self.delta_q = delta_q
        self.Tb_f = tb_f
        self.q_capacity = self.calculate_capacity(tb_f=self.Tb_f)
        self.q = self.q_capacity + self.delta_q
        if self.q_capacity != 0. and self.q_capacity is not None:
            self.soc = self.q / self.q_capacity
        self.resetting = True

    def apply_soc(self, soc, tb_f, delta_q):
        """Memory set, adjust bookkeeping as needed.  delta_q preserved"""
        self.soc = soc
        self.Tb_f = tb_f
        self.q_capacity = self.calculate_capacity(tb_f=tb_f)
        self.q = self.soc * self.q_capacity
        # self.q_eps = delta_q + self.q_capacity * (1. - self.soc)
        # self.delta_q = -self.q_capacity * (1. - self.soc) + self.q_eps
        self.delta_q = delta_q
        self.resetting = True  # momentarily turn off saturation check

    def calculate_capacity(self, tb_f):
        """Capacity"""
        try:
            res = self.q_cap_rated_scaled * (1. + self.chemistry.dqdt * (tb_f - self.chemistry.rated_temp))
            if self.reset:
                pass
            else:
                pass
        except IOError:
            res = 1
        return res

    def count_coulombs(self, OPT, chem, dt, reset, tb_f, charge_curr, sat, saturated):
        """Count coulombs based on true=actual capacity
        Inputs:
            dt              Integration step, s
            tb_f            Battery temperature, deg C  (filtered usually to reduce electrical noise artifacts)
            charge_curr     Charge, A
            sat             Indicator that battery is saturated (VOC>threshold(temp)), T/F
            coul_eff        Coulombic efficiency - the fraction of charging input that gets turned into usable Coulombs
            use_mon_soc     Command to drive integrator with input mon_soc
            soc             Auxiliary integrator setting, fraction soc
        """
        if self.chm != chem:
            self.chemistry.assign_all_mod(chem)
            self.chm = chem
        self.reset = reset
        self.d_delta_q = charge_curr * dt
        if charge_curr > 0. and not self.tweak_test:
            self.d_delta_q *= self.chemistry.coul_eff
        self.sat = sat
        self.saturated = saturated
        self.Tb_f = tb_f

        # if charge_curr < 0.:
        #     print(f"{OPT.mon_run.time[G.i]=} {OPT.mon_run.time[G.i]-OPT.mon_run.time[G.i-1]=} {OPT.mon_run.dt[G.i]=} {dt} {OPT.mon_run.ib_charge[G.i]=} {charge_curr} {OPT.mon_run.ib_charge[G.i] * OPT.mon_run.dt[G.i]=} {dt * charge_curr} {OPT.mon_run.d_delta_q[G.i]=} {self.d_delta_q}")

        # Saturation.   Goal is to set q_capacity and hold it so remembers last saturation status.
        if self.saturated:
            if self.d_delta_q > 0:
                self.d_delta_q = 0.
                if ~self.resetting:
                    self.delta_q = 0.
        self.resetting = False  # one pass flag.  Saturation debounce should reset next pass

        # Integration
        self.q_capacity = self.calculate_capacity(tb_f=self.Tb_f)
        if OPT.use_mon_soc:
            self.soc = OPT.mon_run.soc[G.i]
            self.q = self.q_capacity * self.soc
            self.delta_q = self.q - self.q_capacity
        else:
            if not self.resetting and not self.reset:
                # self.delta_q = max(min(self.delta_q + d_delta_q - self.chemistry.dqdt*self.q_capacity*tb_f_rate*dt,
                #                        0.0), -self.q_capacity*1.5)
                #  Because delta_q is off of saturation and capacity book-kept elsewhere for soc, don't need to book
                #  the temperature effect here
                self.delta_q = max(min(self.delta_q + self.d_delta_q, 0.0), -self.q_capacity*1.5)

                self.q = self.q_capacity + self.delta_q
                if self.delta_q < -100.:
                    pass

        # Normalize
        if self.q_capacity != 0. and self.q_capacity is not None:
            self.soc = self.q / self.q_capacity
        self.soc_min = self.chemistry.lut_min_soc.interp(self.Tb_f)
        self.q_min = self.soc_min * self.q_capacity

        # Save and return
        # print('Mon CC:  charge_curr', charge_curr, 'dt', dt, 'd_delta_q', d_delta_q,'temp_lim', self.temp_lim, 'cap', self.q_capacity)
        return self.soc

    def load(self, delta_q):
        """Load states from retained memory"""
        self.delta_q = delta_q
